SPIFlashLogger.device.lib.nut

```diff
@@ -0,0 +1,847 @@
+            if (!preparing)
+                _sectorsMap[sector].init();
+        }
+        _disable();
+    }
+}
+
+class SPIFlashLogger.LoggerSector {
+    // position for reading
+    _pos_read = 0;
+    // position for writing
+    _pos_write = 0;
+    // start/end address
+    _start = 0;
+    _end = 0;
+    // spi flash object
+    _flash = null;
+    _logger = null;
+    // indicates if we need to erase sector before write
+    _eraseBeforeWrite = false;
+
+    // Default status of sector is free
+    // which is equal to the 0xFFFFFF
+    static SECTOR_STATUS_FREE = 0;
+    // Indicates that we can read metadata from that sector
+    static SECTOR_HAS_METADATA = 1;
+    // Indicates that sector has some written data
+    static SECTOR_WRITE_START = 1 << 1;
+    // Indicates that sector was closed on writing
+    static SECTOR_WRITE_DONE = 1 << 2;
+    // Indicates that sector has at least one start code
+    static SECTOR_HAS_START_CODE = 1 << 3;
+    // indicates that start code could be invalid
+    // because object was erased
+    static SECTOR_HAS_REMOVED_OBJECT = 1 << 4;
+
+    // header bytes with status of current sector
+    _status = 0xFFFF;
+    // chunk map indicates if all chunks are busy
+    // 0 - chunk was written
+    // 1 - chunk is free
+    _chunkMap = 0xFFFF;
+
+    //
+    // Make a flash sector object
+    // @param {integer} - start address of sector
+    // @param {integer} - end address of sector
+    // @param {Flash} - hardware.spiflash object
+    // @param {SPIFlashLogger} - parent class instance
+    constructor(start = 0, end = null, flash = null, logger = null) {
+        _start = start;
+        _end = end;
+        _flash = flash;
+        _logger = logger;
+        _pos_write = SPIFLASHLOGGER_SECTOR_METADATA_SIZE;
+    }
+
+    // read current status of flash's sector
+    function init() {
+        _logger._enable();
+        local meta = _flash.read(_start, SPIFLASHLOGGER_SECTOR_METADATA_SIZE);
+        _logger._disable();
+        // Parse the meta data
+        meta.seek(0);
+        // read first 6 bytes
+        _status = meta.readn('b');
+        _chunkMap = meta.readn('w');
+
+        if (isStoppedWriting()) {
+            _pos_write = SPIFLASHLOGGER_SECTOR_SIZE;
+        } else if (isStartWriting()) {
+            _pos_write = getFreeChunkAddress();
+        } else if (isFree()) {
+            _pos_write = SPIFLASHLOGGER_SECTOR_METADATA_SIZE;
+        }
+    }
+
+    function getFreeSpace() {
+        // there is no free space if sector
+        // was closed for writing
+        if (~_status & SECTOR_WRITE_DONE)
+            return 0;
+        return SPIFLASHLOGGER_SECTOR_SIZE - _pos_write;
+    }
+
+    function _prepareForWrite() {
+        // clean and re-init sector
+        if (!this.isFree())
+            erase();
+    }
+
+    function write(object, object_pos = 0, len = null) {
+        if (len == null)
+            len = object.len();
+
+        if (_eraseBeforeWrite)
```

Comment:
Not sure I understand the logic around `_eraseBeforeWrite`. Can you please elaborate on this?

Reply:
Use case it very strange from my point of view, but it is available in the original code:

logger.setPosition(0); // or i*SECTOR_SIZE
logger.write("SOME DATA");

If user set position which is at the beginning of any sector (or position is in the start code of sector), then sector should be cleaned on write (of course if sector was dirty).